### PR TITLE
Introduce custom OpenSSL BIOs for providing direct buffer read/write

### PR DESCRIPTION
**Pull request by vekterli**
@havardpe please review
@bjorncs FYI

BIOs offer a dynamic view into source or sink (const/mutable) buffers
and avoids overhead of copying from/to memory BIOs. Also strictly
enforces buffer sizes to ensure there are no hidden reallocs.

Additionally make code OpenSSL 1.1+ and TLSv1.3 compatible.

### CODE REVIEW

**Comment by vekterli** [comment]
Travis failure is unrelated; seems to fail early when trying to fetch a docker image

File: vespalib/src/tests/net/tls/direct_buffer_bio/direct_buffer_bio_test.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    BioPtr mutable_bio;
+    BioPtr const_bio;
+    vespalib::string tmp_buf;
+
+    Fixture()
+        : mutable_bio(new_mutable_direct_buffer_bio()),
+          const_bio(new_const_direct_buffer_bio()),
+          tmp_buf('X', 64)
+    {
+        assert(mutable_bio && const_bio);
```
Comment (havardpe) [inline vespalib/src/tests/net/tls/direct_buffer_bio/direct_buffer_bio_test.cpp]: prefer ASSERT_TRUE to normal assert in test fixtures

File: vespalib/src/vespa/vespalib/net/tls/impl/direct_buffer_bio.cpp [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+{
+    WHEN_NO_OPENSSL_BIO_TYPE_BUG(LOG_ASSERT(is_mutable_bio(bio)));
+    set_bio_mutable_buffer_view(bio, &_view);
+}
+
+MutableBufferViewGuard::~MutableBufferViewGuard() {
+    unset_bio_buffer_view(_bio);
+}
+
+}
```
Comment (havardpe) [inline vespalib/src/vespa/vespalib/net/tls/impl/direct_buffer_bio.cpp]: missing newline

**Review by havardpe — APPROVED** [review]

**Comment by vekterli** [comment]
Comments addressed. I'll manually verify that this compiles and tests cleanly on RHEL6 before merging

**Pull request merged**